MoreLinq/DeconstructibleEnumerable.cs

```diff
@@ -0,0 +1,86 @@
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="DeconstructibleEnumerable{T}"/>
+        /// with a sequence.
+        /// </summary>
+        /// <param name="source">
+        /// The source sequence on which to enable deconstruction.</param>
+
+        public DeconstructibleEnumerable(IEnumerable<T> source) =>
+            _source = source ?? throw new ArgumentNullException(nameof(source));
+
+        /// <inheritdoc />
+        public IEnumerator<T> GetEnumerator() =>
+            _source != null ? _source.GetEnumerator()
+                            : throw CreateTooShortError();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        void Deconstruct(int count, out T item1, out T item2, out T item3, out T item4, out T item5, out T item6, out T item7, out T item8, out T item9, out T item10, out T item11, out T item12, out T item13, out T item14, out T item15, out T item16)
+        {
+            Debug.Assert(count >= 2);
+            Debug.Assert(count <= 16);
+
+            using (var e = GetEnumerator())
+            {
+                item1  = e.MoveNext() ? e.Current : throw CreateTooShortError();
+                item2  = e.MoveNext() ? e.Current : throw CreateTooShortError();
+                item3  = count >=  3 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item4  = count >=  4 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item5  = count >=  5 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item6  = count >=  6 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item7  = count >=  7 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item8  = count >=  8 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item9  = count >=  9 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item10 = count >= 10 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item11 = count >= 11 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item12 = count >= 12 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item13 = count >= 13 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item14 = count >= 14 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item15 = count >= 15 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                item16 = count >= 16 ? e.MoveNext() ? e.Current : throw CreateTooShortError() : default;
+                if (e.MoveNext())
+                    throw CreateTooLongError();
+            }
+        }
+
+        static InvalidOperationException CreateTooShortError() =>
+            new InvalidOperationException("Sequence too short.");
```

Comment:
I was just being lazy but I'll go clean up the room now. :wink: Done in e885659c7d9f8bb0e04c79bb898191c58bdf9f70.

Reply:
You didn't print the actual number. Was that intentional?

Reply:
Are you looking at outdated code? The code after the changes in e885659c7d9f8bb0e04c79bb898191c58bdf9f70 should read as follows:

```c#
static InvalidOperationException CreateTooShortError(int expectedCount) =>
    CreateLengthMismatchError(expectedCount, "few");

static InvalidOperationException CreateTooLongError(int expectedCount) =>
    CreateLengthMismatchError(expectedCount, "many");

static InvalidOperationException CreateLengthMismatchError(int expectedCount, string actualQuantifier) =>
    new InvalidOperationException($"Sequence contains too {actualQuantifier} elements when exactly {expectedCount} {(expectedCount == 1 ? "was" : "were")} expected.");
```

Reply:
Yes, I'm looking at the new code. I mean in the short case, we also know the actual number of elements encountered in the sequence.

Reply:
I see what you're saying now. Unfortunately, your comment appears tagged next to the outdated code so that threw me off for a bit.

Reply:
> we also know the actual number of elements encountered in the sequence.

True but wouldn't the information be somewhat superfluous? Does it add any value, especially when we can't do any better in the case of too many elements? Isn't knowing _the exact count_ expected the real thing? Call me lazy but I'd like to argue that this is an improvement we can make post-merge because it doesn't seem seem like a design defect. Also bear in mind that `Fold` has been the same for years now (and yet no one's complained that it could be better):

https://github.com/morelinq/MoreLINQ/blob/d0848d9bf9bfbde2b5141c16e90203d48fa44da6/MoreLinq/Fold.cs#L96-L99